tahoe_sites/backends.py

```diff
@@ -0,0 +1,62 @@
+"""
+Authentication backends for Tahoe's multi-tenancy.
+"""
+from django.conf import settings
+from django.contrib.auth.backends import AllowAllUsersModelBackend, ModelBackend
+
+from tahoe_sites.api import get_current_site, get_organization_by_site, get_organizations_for_user
+
+
+class DefaultSiteBackend(ModelBackend):
+    """
+    User can log in to the default/root site (edx.appsembler.com) because it is required during the signup.
+    Also, superusers (appsembler admins) can log into any site.
+    """
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        """
+        Authenticate superusers only.
+        """
+        user = super().authenticate(request, username, password, **kwargs)
+
+        if user:
+            if user.is_superuser or get_current_site().id == settings.SITE_ID:
+                return user
+        return None
+
+
+class OrganizationMemberBackend(AllowAllUsersModelBackend):
+    """Backend for organization based authentication
+
+    This class is an extension of Django's `AllowAllUserModelBackend`
+
+    This class checks that the user to authenticate belongs to one of the
+    organizations in the specified site
+
+    This class extends `AllowAllUserModelBackend` instead of `ModelBackend`
+    because users need to be able to authenticate when the `user.is_active` is
+    `False`. The reason for this is that Open edX has an email verification
+    scheme that uses `User.is_active` in order to prevent user activity until
+    the users have verified their email. Effectively, edx-platform LMS user
+    authentication state is used to manage authorization state.
+
+    The key problem is that Django's `django.contrib.auth.backends.ModelBackend`
+    is called with `not user.is_active` and Ironwood introduced a check to test
+    authentication on a not yet authorized user. This breaks Tahoe multisite
+    behavior. Extending `AllowAllUsersModelBackend` restores correct behavior.
+
+    For further reference, see settings files and read Django documentation for
+    `AUTHENTICATION_BACKENDS`
+    """
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        """
+        Authenticate organization learners.
+        """
+        user = super().authenticate(request, username, password, **kwargs)
+
+        site = get_current_site()
+        is_default_site = site.id == settings.SITE_ID
+        if not is_default_site and user and not user.is_superuser:
```

Comment:
**Required change:** I think we should put this logic into an api helper instead of re-implementing it every time. It makes clear that this is a supported check and we maintain it.

```suggestion
        if not is_default_site(site):
```

Reply:
`is_main_site` is added. Please check the test named `test_superuser` which complies with the original backend logic. Does it make sense to disallow super users in this backend?

Reply:
Thank you @OmarIthawi

Reply:
> `is_main_site` is added. Please check the test named `test_superuser` which complies with the original backend logic. Does it make sense to disallow super users in this backend?

Yes, this is intentional. Thanks.